### PR TITLE
Update r-base to 3.4.0.

### PR DESCRIPTION
**Pull request by johanneskoester**
Along with this, I tried to remove all patches, in the hope that they are not needed anymore. This will improve maintainability in the future. At least on my machine, it seems to work fine, let's see what CI says.

### Needed patches
* [x] 0002-Win32-Fix-trio-config-h-include-depth-issue.patch ([reason](https://ci.appveyor.com/project/conda-forge/r-base-feedstock/build/1.0.30#L1403))
* [x] 0007-Win32-MSYS2-ENV_CONV_EXCL_R_ARCH.patch
* [x] 0003-Win32-Do-not-static-libgcc-link.patch

### CODE REVIEW

**Comment by ocefpaf** [comment]
> Along with this, I tried to remove all patches, in the hope that they are not needed anymore.

@mingwandroid if you have some spare time can you take a look at the patches? I believe you created most of them.

> This will improve maintainability in the future.

:+1: 

> At least on my machine, it seems to work fine, let's see what CI says.

What platform is that? Linux?

@johanneskoester what is the strategy here? Drop 3.3.2 and build only for 3.4?

@vinjana is requesting 3.3.1 in https://github.com/conda-forge/r-reshape-feedstock/pull/1#discussion_r115927988 but I am not sure the CIs can handle 3 items in the matrix. (Not to mention the human support fot 3 R versions.)

My biased view is that conda-forge should support only 3.4. But I am not a heavy R users!

**Comment by johanneskoester** [comment]
@ocefpaf, indeed, I only want to support 1 version (but leave any old versions that have been previously built on anaconda for reproducibility). Further, we figured it is sufficiently secure to pin against 3.4 instead of 3.4.x. So once this works, CONDA_R has to become "3.4" in the depending feedstocks.
This also means that no rebuilds of downstream packages have to be made until R 3.5 is released (maybe in a year).

**Comment by johanneskoester** [comment]
@bgruening I have added you as a maintainer if that is ok?

**Comment by ocefpaf** [comment]
> @ocefpaf, indeed, I only want to support 1 version

I guess these are bad news for @vinjana.

> (but leave any old versions that have been previously built on anaconda for reproducibility).

Oh yes. I don't see the point of removing those builds!

> Further, we figured it is sufficiently secure to pin against 3.4 instead of 3.4.x. So once this works, CONDA_R has to become "3.4" in the depending feedstocks.
This also means that no rebuilds of downstream packages have to be made until R 3.5 is released (maybe in a year).

:+1:

**Comment by johanneskoester** [comment]
To clarify, I am totally fine with somebody else supporting more than the one version. But I simply don't have the resources to do that.

**Comment by ocefpaf** [comment]
> To clarify, I am totally fine with somebody else supporting more than the one version. But I simply don't have the resources to do that.

That answer the human part of the support. (I also don't have time for that).
But there is still the strain on the CIs...

**Comment by johanneskoester** [comment]
And of course, everything except windows works fine again...

**Comment by johanneskoester** [comment]
Btw. what is the strategy of windows support for conda-forge? We could proceed much faster if windows was not needed... At least in my field, windows is not used at all, at least not for data science with R.

**Comment by ocefpaf** [comment]
> Btw. what is the strategy of windows support for conda-forge? We could proceed much faster if windows was not needed... At least in my field, windows is not used at all, at least not for data science with R.

We cannot ignore Windows entirely but I do understand how hard it is to support it. Before we merge this skipping Windows let's try to:

1. figure out if any of those patches needs refreshing
2. wait for a comment from @mingwandroid who has a vast experience with R/MSYS2

After that I guess there is not much to do beyond the usual skip win, create an issue, address it in the future :weary:

**Comment by johanneskoester** [comment]
Yeah, don't worry. I also don't want skip windows now that we have it already. It is just frustrating :-).

**Comment by mingwandroid** [comment]
You should definitely not disable any of my patches without a full understanding of what they do and why they are no longer needed (i.e. you are completely sure they've been merged upstream or some equivalent fix has be applied).

> - 0001-Win32-Stop-confusing-PECOFF_SHLIBS-with-CYGWIN_BUILD.patch

This one is actually WIP, eventually I want to use Autotools to build R-base on Windows and this is work towards that goal. This one could be disabled but it's harmless.

> - 0002-Win32-Fix-trio-config-h-include-depth-issue.patch

Both necessary and obvious.

> - 0003-Win32-Do-not-static-libgcc-link.patch 

Without this one DLL unloading will cause crashes. It is only appropriate for builds of R where DLLs never load other DLLs which is not the approach we have gone for (because it defeats the purpose of building a distro out of shared libraries).

> - 0007-Win32-MSYS2-ENV_CONV_EXCL_R_ARCH.patch

Without this patch, R cannot build any of its packages that require calling `configure` (about half of the ones that require compilation, maybe more).

**Comment by mingwandroid** [comment]
BTW, I don't have time to build R 3.4.0 on defaults just now. I also want to rebuild all of R on Linux using GCC 6.3.0 or perhaps even 7 using the new CI system @msarahan has been working on.

I also find the .0 releases pretty buggy (in general and for R) so I actually think waiting about a month after a .0 release to upgrade is a very sensible course of action.

**Comment by johanneskoester** [comment]
@mingwandroid thanks for your insights about the patches. I think what we can do here is to use the minimum set of patches to build R successfully and then to re-add them whenever we encounter a problem during a rebuild of the feedstocks. 0002, 0003 and 0007 I can activate right away.

**Comment by johanneskoester** [comment]
Since we won't need to rebuild everything with each new patch version update of r-base, I think it should be fine to start with 3.4.0.

**Comment by mingwandroid** [comment]
@johanneskoester, the patches were not written because I had some free time. You should take the opposite approach and assume that they are all necessary.

**Comment by mingwandroid** [comment]
> Since we won't need to rebuild everything with each new patch version update of r-base

What makes you say that? I've found that the be incorrect.

**Comment by johanneskoester** [comment]
Well, this is what @bgruening was told when asking various colleagues. They meant it is usually not problematic to use a package built for a different patch version. If you had a different experience, can you point us to the exact problem that occurred?

**Comment by johanneskoester** [comment]
Regarding the patches: I guess it is obvious that nobody assumes you produced them just for fun. However, we have a new version of R here, so I just wanted to give it a try. Of course, if you would be so kind to let us know about the purpose of all patches, I am very happy to include them again. I just want to know why before doing the work.

**Comment by mingwandroid** [comment]
Currently conda-build pins exactly to the r-base version (though yes conda-build 3 allows for control over that).

Knowing how DLLs and ABIs work and how the R project is organised gives me very strong evidence that you will have a very bad time if you try mix these things.

**Comment by johanneskoester** [comment]
So, you mean that R breaks the ABI in patch versions? We have considered that and not found evidence for that in the changelogs, but maybe I missed it. Can you point us to such a change? If that is really the case, we could still pin to 3.4 for non-compiled packages...

**Comment by mingwandroid** [comment]
Can you point to where R claims to follow static versioning? It is my impression that X.X.X is just their own rough feeling on degree of change.

Even if they do, changing configuration options sinks sematic versioning.

**Comment by mingwandroid** [comment]
Sematic versioning. On a phone..

**Comment by mingwandroid** [comment]
Argh. Semantic..

**Comment by johanneskoester** [comment]
:-). I can't because they don't have such a statement. However, besides my personal and the experience of others, there is some evidence around that they won't break things in a patch release:

1. In the sources (base/R/library.R), they have an error if a package was build before 3.0.0, and a warning if a package was built with a **newer** version than the current. No warning in case of an older version.
2. The function `update.packages()` from r-base, which is recommended to be executed after updating R and copying over packages to the new version (as e.g. done by [installr](https://github.com/talgalili/installr/)), has an option checkBuilt, which, when set to true rebuilds packages if and only if the package has been built under an older major.minor version. Not under a changed patch version.

Hence, even without an official statement, the facilities in the R code show that major.minor updates are the thing to worry about, not patch versions.

File: recipe/meta.yaml [hunk 1]
```diff
@@ -65,41 +65,41 @@ requirements:
...
-    - {{native}}bwidget        # [win]
-    - {{native}}tktable        # [win]
+    - {{native}}libjpeg-turbo 1.4.*  # [win]
+    - {{native}}libiconv 1.14        # [win]
+    - {{native}}gmp 6.1.*            # [win]
+    - {{native}}fftw 3.3.*           # [win]
+    - {{native}}mpfr 3.1.*           # [win]
+    - {{native}}libsndfile 1.0.*     # [win]
+    - {{native}}bwidget 1.9.*        # [win]
+    - {{native}}tktable 2.10         # [win]
```
Comment (mingwandroid) [inline recipe/meta.yaml]: This pinning will prevent using that version of R when I update the MSYS2 packages. This is far too rigid.

Comment (johanneskoester) [inline recipe/meta.yaml]: Wouldn't the old versions still be avaialable? If these are dynamically linked libraries (which I don't know), we can indeed relax it again. Basically it was a try to react on your concern that statically linked libraries could impact the ABI of R when doing an update.

**Comment by mjsteinbaugh** [comment]
Hi all, any update on this? I have a growing list of Bioconductor packages that require R 3.4 on our HPC cluster.

**Comment by johanneskoester** [comment]
I think we came to the conclusion that it is too risky to change the pinning. In turn there will be an update to 3.4.1 or 3.4.2 only (due to time and space restricitions we cannot rebuilt all r packages for every single point release). It has yet to be decided which point release to take.

**Comment by chapmanb** [comment]
Johannes;
It looks like 3.4.1 is scheduled for tomorrow, do you think that's a reasonable target to move to 3.4 compatibility?

**Comment by mingwandroid** [comment]
It sounds like 3.4.1-patched might be [worth](https://stat.ethz.ch/pipermail/r-devel/2017-June/074552.html) waiting for instead.

**Comment by mjsteinbaugh** [comment]
Compiling against 3.4.1 should work great, for at least a year. The shift from R 3.3 to 3.4 is significant, and [Bioconductor 3.5](http://www.bioconductor.org/news/bioc_3_5_release/) requires R 3.4 for all future package updates.

**Comment by johanneskoester** [comment]
Waiting for 3.4.1-patched sounds good to me. Feel free to ping me here once it becomes available.

**Comment by johanneskoester** [comment]
Closing for now. There will be a followup PR that simply migrates the new 3.4.x recipe from conda/conda-recipes.

**Comment by bgruening** [comment]
@johanneskoester are you going to create the r-base package? If not I might have a look over the weekend.

**Comment by johanneskoester** [comment]
That would be great. I won't be able to work on this until some time in December. In principle it should be sufficient to simply copy over the new recipe from conda/conda-recipes/r-base and adjust the pinnings to match conda-forge.

**Pull request closed without merging**